### PR TITLE
Adds a --chmod flag to ADD and COPY

### LINKED ISSUE #34819: Proposal: Add --chmod flag to ADD/COPY commands (analogous to --chown)
**Description**
I've bumped into #34263 being merged and it hit me that `--chmod` would be nice in conjunction with this feature.

I frequently build Linux images from a Windows machine and I'm kind of annoyed with the 755 mode the added/copied files end up with.

ADD/COPY with `--chown` and `--chmod` together will be a readable and maintainable notation delivering precise control over permissions.

I'm aware of #29853, but I don't think this issue is a duplicate (that one seems to be stale).

### PR DESCRIPTION
**Pull request by Ahmadposten**
Closes https://github.com/moby/moby/issues/34819 

**What I did**: Implemented a `--chmod` flag for `ADD` and `COPY` instructions to allow the user to set file permissions in the same time while adding or copying without having to explicitly run `chmod` command.

**How I did it**: I implemented it in a way analogous to how `--chown` was implemented where `fixPermissions` sets the file permissions if `--chmod` was provided.

**How to verify it**: Either by running the tests or by manually adding `--chmod=xxx` to one of your docker files and inspecting the resulting image. 

**Summary**
This adds the ability to change file permissions while copying or adding by supplying an additional optional flag `--chmod`
For example: `COPY --chmod=777 . /new_dir` will copy files to `/new_dir` with open permissions of `777`

It also supports symbolic modes 

For example: `COPY --chmod u+rwx,go+rx ./new_dir` will result in `755` permission

### CODE REVIEW

**Comment by Ahmadposten** [comment]
I see that pipeline `windowsRS1 ` check seems to be failing for some reason that looks unrelated to this PR changes.
If it is related can someone please clarify

**Comment by Ahmadposten** [comment]
ping @vdemeester

**Comment by glensc** [comment]
in my opinion `chmod bits` is bad, promotes use of bad practices, i.e `0666` and `0777` permissions. as the copy is recursive, there must be way to specify different permissions for files and directories.

unix chmod supports ugoa and rwx symbolic modes: https://en.wikipedia.org/wiki/Chmod#Symbolic_modes

for single value like numeric bits, umask suits better, but it's usually harder to understand than just permission bits.

**Comment by Ahmadposten** [comment]
@glensc most people are familiar with `chmod bits` and rarely use symbolic modes.
May be both needs to be supported.
As for the recursive part are you proposing that the `chmod` flag to act in a different way than copy like the default would be to change mode of the directory/file (not recursively) and allow a `-R` to make it act recursively? 

On another note: I am not familiar with how things are done in the docker project, if we support both do we need to include the `symbolic mode` into this PR or start with what we have now and create a different issue/PR to address that as improvement ?

**Comment by reporter123** [comment]
Symbolic mode isn't just more readable its the only way to add a flag without having to worry about every other flag too. Ideally both should be provided since chmod uses both forms. If we can only do one the symbolic version is far more flexible than bits mode. Think about add an exec flag to a file. chmod 3333 forces just the exec flag to be on. chmod a+x adds the exec flag without having to force the other flags as well. This can be factor with docker builds where only one flag is causing and issue. Someone just post an example of this scenario in #34819.

**Comment by Ahmadposten** [comment]
Will this PR ever be reviewed?

**Comment by mhumeSF** [comment]
I don't think golang supports symbolic mode natively. So I think a parser would have to be written to add support. With that in mind, I think this would be a great improvement and would work for a majority of use cases.

File: builder/dockerfile/internals_linux.go [hunk 1]
```diff
@@ -49,6 +49,15 @@ func parseChownFlag(chown, ctrRootPath string, idMappings *idtools.IDMappings) (
 	return chownPair, nil
 }
 
+func parseChmodFlag(chmodStr string) (uint16, error) {
```
Comment (c0b) [inline builder/dockerfile/internals_linux.go]: needs to parse symbolic names, should mimic most `chmod` flags like `a+rX,ug+w` if not all of chmod.

**Comment by c0b** [comment]
> @Ahmadposten commented on Feb 17
> @glensc most people are familiar with chmod bits and rarely use symbolic modes.

are you saying for yourself ? or how are you knowing most people?  from Admin / OPS view I was taught symbolic modes are more accurate to add or remove the exact permission bits intended and not touching other bits,  and from my programming  experience in C, I do always use those pre-defined S_IRUSR / S_IWUSR / S_IXGRP symbolic names than the numeric ones
those are the best practice from both Admin and Linux C programming  and I don't think I am the only one

I do want this `--chmod` in ADD and COPY, but before it's eventually going be merged, I think the parseChmodFlag needs to parse symbolic names, should mimic most chmod flags like a+rX,ug+w if not all flags of chmod.

**Comment by Ahmadposten** [comment]
@c0b Definitely not saying for myself but from experience with people who write docker files 

I don't disagree that this is needed and I don't mind implementing it in this PR

**Comment by mhumeSF** [comment]
@c0b I think also doing a search for "filename:Dockerfile chmod" shows people trying to add the executable permission in some form or fashion of `+x`, `a+x`, `o+x`, or `777`.

File: builder/dockerfile/internals_linux.go (hunk 1 above)
Comment (Ahmadposten) [reply builder/dockerfile/internals_linux.go]: Done

**Comment by Ahmadposten** [comment]
I implemented symbolic modes chmod.

Unfortunately golang `os.Chmod` does not allow for symbolic mode

I had to implement a mini-parser that converts symbols to a `uint` to pass it to `os.Chmod`

cc: @c0b

**Comment by Ahmadposten** [comment]
It looks like there is no interest in this so I close it

**Comment by c0b** [comment]
sorry, didn't notice this earlier but  `21 days ago`  I don't think is good enough time to declare no interest to close

**Comment by Ahmadposten** [comment]
@c0b 
I think 6 months is quite enough

**Comment by reporter123** [comment]
@Ahmadposten it would be if there had been communication whatever from upsteam. You closed this 21 days after the last update to patch. I cann't make you change your mind but know I will be opening a new PR with this change if you no longer want to wait. Full credit will given for the work done.

**Comment by Ahmadposten** [comment]
@reporter123 I see 7 people assigned as reviewers and for months i've been pinging them to take a look or atleast give any feedback even if this feedback was "this is useless, we don't want it" instead I got nothing
When I got a piece of feedback finally I addressed it and waited for 21 days 
and I can see that no one noticed that I did any change so it would probably have taken another 6 months to get another response.

This is not normal even for this projects I can see that PRs get reviewed and merged a lot faster than this 

I don't know what is causing this

**Comment by Ahmadposten** [comment]
I was actually interested in being a part of this project but what is happening here is not encouraging  at all

**Comment by douglasmiranda** [comment]
@reporter123 Please, do. :)

**Comment by AkihiroSuda** [comment]
Implemented in moby/buildkit#1492

**Comment by iraklisg** [comment]
In my Debian system using docker engine `5:19.03.12~3-0~debian-buster` I get the following error:
```
Dockerfile parse error line xx: Unknown flag: chmod
```
Is this new feature going to be available on a latest version of docker?

**Comment by AkihiroSuda** [comment]
@tonistiigi Can we have `dockerfile/1.2.0` ahead of Docker 20.03 so that 19.03 users can use the instruction?

**Comment by tonistiigi** [comment]
@AkihiroSuda yes

**Comment by hosnimed** [comment]
> In my Debian system using docker engine `5:19.03.12~3-0~debian-buster` I get the following error:
> 
> ```
> Dockerfile parse error line xx: Unknown flag: chmod
> ```
> 
> Is this new feature going to be available on a latest version of docker?

Same problem on Docke for mac with exprimental features enabled

`COPY --from=app --chmod=755 opt /opt`



> Client: Docker Engine - Community
 Version:           19.03.12
 API version:       1.40
 Go version:        go1.13.10
 Git commit:        48a66213fe
 Built:             Mon Jun 22 15:41:33 2020
 OS/Arch:           darwin/amd64
 Experimental:      true

> Server: Docker Engine - Community
 Engine:
  Version:          19.03.12
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.13.10
  Git commit:       48a66213fe
  Built:            Mon Jun 22 15:49:27 2020
  OS/Arch:          linux/amd64
  Experimental:     true
 containerd:
  Version:          v1.2.13
  GitCommit:        7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683

**Pull request closed without merging**